### PR TITLE
Add whitelist function

### PR DESCRIPTION
**Pull request by phamminh0811**

### CODE REVIEW

File: ido/src/modules/testing.rs [hunk 1]
```diff
@@ -3,18 +3,22 @@ use crate::*;
 #[near_bindgen]
 impl IDOContract{
 
-    fn internal_new_project_1()-> Project{
-        panic!("The function has not been implemented yet.");
-    }
+    
 
     pub fn create_sample_projects(&mut self) {
-        self.assert_test_mode_and_owner();
```
Comment (tran-thanh-phong) [inline ido/src/modules/testing.rs:28]: Why do we remove the owner validation here?

File: ido/src/modules/project.rs [hunk 2]
```diff
@@ -244,23 +244,43 @@ impl IDOContract {
...
+                self.internal_register_fixed_xtoken_project(project_id,account_id,xtoken);
+            },
+            WhitelistType::Ticket => {
+                self.internal_add_account(&account_id, project_id);
+            }
+        }
+    }
+
+    pub(crate) fn internal_add_account(&mut self, account_id: &AccountId, project_id: ProjectId){
+
```
Comment (tran-thanh-phong) [inline ido/src/modules/project.rs:342]: Please don't leave an empty line like this.

File: ido/src/modules/project.rs [hunk 3]
```diff
@@ -244,23 +244,43 @@ impl IDOContract {
...
+        let mut projects_by_account = self.internal_get_projects_by_account_or_default(account_id);
 
         assert!(!projects_by_account.contains(&project_id),"Already register whitelist this project");
         projects_by_account.insert(&project_id);
-        self.projects_by_account.insert(&account_id,&projects_by_account);
+        self.projects_by_account.insert(account_id,&projects_by_account);
```
Comment (tran-thanh-phong) [inline ido/src/modules/project.rs]: Please leave a space after a comma (,), the code will look nice.

File: ido/src/modules/project.rs [hunk 4]
```diff
@@ -276,4 +296,15 @@ impl IDOContract {
         let project = self.internal_get_project_or_panic(project_id);
         
     }
+
+    pub(crate) fn internal_get_required_xtoken(&self, project_id: ProjectId) -> U128{
```
Comment (tran-thanh-phong) [inline ido/src/modules/project.rs]: The _get_project_ function returns the project information that includes the number of required xTokens. We should reuse it and don't need to add a new function. This function should be removed.

File: ido/src/modules/project_internal.rs [hunk 5]
```diff
@@ -0,0 +1,349 @@
+use crate::*;
```
Comment (tran-thanh-phong) [inline ido/src/modules/project_internal.rs]: This file has been removed and the code was moved to the testing.rs file. Please sync code from develop branch.

File: ido/src/modules/xtoken.rs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    pub (crate) fn proccess_register_whitelist(&mut self, account_id: AccountId, project_id: ProjectId, point: U64, xtoken: u128) -> bool{
+        env::log(format!("proccess_register_whitelist(account_id: {}, project_id: {}, point: {})", account_id, project_id,point.0).as_bytes());
+
+        assert!((point.0 as u128) >= xtoken,"User don't have enough XToken");
+
+        self.internal_add_account(&account_id, project_id);
+        true
+
+
```
Comment (tran-thanh-phong) [inline ido/src/modules/xtoken.rs]: Please remove these unnecessary empty lines.

File: ido/src/sim_test/ido_test/whitelist_test.rs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+pub fn test_join_whitelist(){
+    let (root,alice,ido,ft_contract,staking_contract,ido_contract) = init();
+    
+    let is_whitelisted : bool = root.view(
+        ido_contract.account_id(),
+        "is_whitelist",
+        &json!({
+            "project_id" : 1
+        }).to_string().as_bytes(),
+    ); // nếu lỗi thì thử .unwrap_json()
```
Comment (tran-thanh-phong) [inline ido/src/sim_test/ido_test/whitelist_test.rs]: It shouldn't use Vietnamese in the source code.

File: ido/src/sim_test/ido_test/whitelist_test.rs [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            "msg": ""
+        }).to_string().as_bytes(), 
+        DEFAULT_GAS, 
+        1
+    );
+
+    // Lock token
+    alice.call(
+        staking_contract.account_id(),
+        "internal_lock",
```
Comment (tran-thanh-phong) [inline ido/src/sim_test/ido_test/whitelist_test.rs]: Should use _lock_ function instead _internal_lock_.

**Review by tran-thanh-phong — COMMENTED** [review]
It looks good! Some comments:
 - Need to sync source code from develop
 - Fix general code format like: unnecessary empty lines, leave a space after a comma, ...
 - Add guidelines on how to call the related functions of the contract by making a new markdown file

File: ido/src/modules/project.rs (hunk 4 above)
Comment (phamminh0811) [reply ido/src/modules/project.rs]: i need to use this function for simulation test. without it, i need to redefine struct Project on simulation test file, which make the code become very complicate.

File: ido/src/modules/testing.rs (hunk 1 above)
Comment (phamminh0811) [reply ido/src/modules/testing.rs:28]: Fixed

File: ido/src/modules/xtoken.rs (hunk 6 above)
Comment (phamminh0811) [reply ido/src/modules/xtoken.rs]: Fixed

File: ido/src/sim_test/ido_test/whitelist_test.rs (hunk 7 above)
Comment (phamminh0811) [reply ido/src/sim_test/ido_test/whitelist_test.rs]: Fixed

File: ido/src/modules/project.rs (hunk 4 above)
Comment (tran-thanh-phong) [reply ido/src/modules/project.rs]: You can import these types from the library using the command:
`use ido::modules::project::ProjectJson;`

**Review by tran-thanh-phong — COMMENTED** [review]
It's better now! I'm happy to merge with two suggestions:
- We want to push all code related to testing process into _/modules/testing.rs_ in order to easy manage. So please move the content of the _modules/project_internal.rs_ file to the _testing.rs_.
- Should rerun the simulation tests because they are running failed.

**Pull request merged**